### PR TITLE
Support relaunch debug protocol

### PR DESCRIPTION
**Pull request by Christian-Holbrook**
Persist the debug session for debug protocol. This is accomplished by parsing the telnet logs for the line "Waiting for debugging connection" and then creating the control port. When the application is closed, we continue to parse the telnet logs for that line. When that line is detected the control port will be created again.

A new signal `app-ready` is emitted to let `BrightScriptDebugSession` know the debug adapters are ready. Both adapters have to emit this signal.

When the app is closed via a home key press, reset the breakpoint managers state for the debug protocol.

### CODE REVIEW

File: src/debugProtocol/client/DebugProtocolClient.ts [hunk 1]
```diff
@@ -253,6 +254,7 @@ export class DebugProtocolClient {
             client: this,
             server: connection
         });
+        await this.emit('control-connected');
```
Comment (TwitchBronBron) [inline src/debugProtocol/client/DebugProtocolClient.ts:257]:
```suggestion
        await this.emit('control-socket-connected');
```

File: src/adapters/DebugProtocolAdapter.ts [hunk 2]
```diff
@@ -389,6 +405,19 @@ export class DebugProtocolAdapter {
         return deferred.promise;
     }
 
+    private async processUnhandedOutputForDebugPrompt(responseText: string) {
```
Comment (TwitchBronBron) [inline src/adapters/DebugProtocolAdapter.ts:408]:
```suggestion
    private async findWaitForDebuggerPrompt(responseText: string) {
```

File: src/adapters/DebugProtocolAdapter.ts [hunk 3]
```diff
@@ -376,6 +389,9 @@ export class DebugProtocolAdapter {
                         lastPartialLine = '';
                     }
                     // Emit the completed io string.
+                    this.processUnhandedOutputForDebugPrompt(responseText.trim()).catch((e) => {
```
Comment (TwitchBronBron) [inline src/adapters/DebugProtocolAdapter.ts:392]: We don't need to .catch this. The whole flow can happen synchronously.

**Review by TwitchBronBron — CHANGES_REQUESTED** [review]
Fe

File: src/adapters/DebugProtocolAdapter.ts [hunk 4]
```diff
@@ -253,6 +263,7 @@ export class DebugProtocolAdapter {
 
             // Listen for the app exit event
             this.socketDebugger.on('app-exit', () => {
+                this.breakpointManager.clearBreakpointLastState();
```
[the remark this answers is not in the document]
Comment (Christian-Holbrook) [reply src/adapters/DebugProtocolAdapter.ts:266]: Done

File: src/adapters/DebugProtocolAdapter.ts (hunk 3 above)
Comment (Christian-Holbrook) [reply src/adapters/DebugProtocolAdapter.ts:392]: Done

File: src/adapters/DebugProtocolAdapter.ts [hunk 5]
```diff
@@ -389,6 +405,19 @@ export class DebugProtocolAdapter {
...
+        for (const line of lines) {
+            if (/Waiting for debugger on \d+\.\d+\.\d+\.\d+:8081/g.exec(line)) {
+                await this.connectSocketDebugger();
+            }
+        }
+    }
+
+    private async connectSocketDebugger() {
+        await this.emit('waiting-for-debugger');
+    }
```
[the remark this answers is not in the document]
Comment (Christian-Holbrook) [reply src/adapters/DebugProtocolAdapter.ts:419]: Done

**Review by TwitchBronBron — APPROVED** [review]
I tested this, everything seems to be working properly. Thanks!

**Pull request merged**